tsconfig.json

```diff
@@ -0,0 +1,11 @@
+{
+  "exclude": ["src"],
+  "references": [{ "path": "./tsconfig.build.json" }],
+  "compilerOptions": {
```

Comment:
I think we should [`"extends": ["@tsconfig/strictest"]`](https://github.com/tsconfig/bases/blob/main/bases/strictest.json) as it adds some nice checks (or we can just pull out those rules from that file and put them in ours). however we could do that in a future PR.

Reply:
I agree that some rules are better covered by eslint (because eslint is more flexible in that you can disable rules per-line, while typescript cannot). however I still think that:

- `isolatedModules` I think it's recommended to enable this in order to enforce writing typescript in a style that is compatible with external tools like `esbuild`, `tsx` etc. [tsx](https://github.com/privatenumber/tsx/blob/7f8a051947ad2af22d8d8f1b99f26ca11ed282d2/docs/typescript.md?plain=1#L62) even recommends enabling it.
- `esModuleInterop`: from my research it's generally recommended to set this to `true`, and `tsx` also recommends it.
- `skipLibCheck` recommended by TypeScript to set to `true`
- `noImplicitOverride` which eslint rule covers this?
- `noImplicitReturns` eslint has consistent-return, however [they recommend that we instead use the typescript rule](https://typescript-eslint.io/rules/consistent-return/).

what's the reason you really dislike these typescript rules?

Reply:
- `isolatedModules` is needed when compiling the TypeScript with tools other than `tsc`. `vitest` is also one such tools, so we should probably enable it indeed. Better would be to use `verbatimModuleSyntax`, but that’s not supported with Vitest apparently.
- `esModuleInterop` is implied by `"module": "node16"`. We don’t need to define it explicitly. Note that this option is risky for libraries (https://github.com/microsoft/TypeScript/issues/52779). The solution is to use `verbatimModuleSyntax`, which we can’t because of Vitest.
- `skipLibCheck` is basically a huge `@ts-ignore` applied to all `.d.ts` files. This is useful if you have dependencies that have type errors. I prefer to disable this, as that helps identify possible compatibility issues with other dependencies. Whether or not to enable this, is subjective.
- `noImplicitOverride` requires TypeScript specific syntax for overriding methods. I slightly prefer to not rely on that. In my experience it’s barely ever relevant though. I haven’t given it that much thought. I could see its value.
- `noImplicitReturns` is a stylistic choice I just don’t like. I pretty strongly prefer the exact opposite style ([`unicorn/no-useless-undefined`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-useless-undefined.md)). I combine this with explicit type annotations ([`@typescript-eslint/explicit-function-return-type`](https://typescript-eslint.io/rules/explicit-function-return-type)). That’s a good idea for libraries anyway, as it gives more control over the types you publish.

Reply:
> * `esModuleInterop` is implied by `"module": "node16"`.

ok, let's remove it.

> - `skipLibCheck: false`

I had problems with it before, where some library had type errors and i couldn't compile because of that, and i believe it's slower. but if it works with `skipLibCheck: false` now, then i guess why not. feel free to set it to false.

> * `noImplicitOverride` requires TypeScript specific syntax for overriding methods. I slightly prefer to not rely on that. In my experience it’s barely ever relevant though. I haven’t given it that much thought. I could see its value.

I also don't rely on this rule very often because i rarely use classes / oo & inheritance in javascript, but when i do, i think it has benefits because I can get an error if some library that I depend on change their signature of an inherited class prop without me knowing. and typescript already has a bunch of other non-js compatible syntax so i don't see any drawback with that.

> * `noImplicitReturns` is a stylistic choice I just don’t like. I pretty strongly prefer the exact opposite style ([`unicorn/no-useless-undefined`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-useless-undefined.md)). I combine this with explicit type annotations ([`@typescript-eslint/explicit-function-return-type`](https://typescript-eslint.io/rules/explicit-function-return-type)). That’s a good idea for libraries anyway, as it gives more control over the types you publish.

Fel free to remove the rule. As for [@typescript-eslint/explicit-function-return-type](https://typescript-eslint.io/rules/explicit-function-return-type) i'm not sure how to make it only apply to public functions, because i don't like having to manually type every internal function

Reply:
I applied this feedback and also applied the new options to `tsconfig.json`.

I also removed `noPropertyAccessFromIndexSignature`, as it’s a stylistic choice our code currently doesn’t adhere to.